### ISSUE TITLE
Enhancements to find and replace

### ISSUE THREAD

**Issue by kirbyfan64** [howl-editor/howl#113]
So, this is a random idea I've had. Basically:

1. Allow using `ctrl_f` and `ctrl_b` in `buffer-replace` like in `buffer-search-forward`. That way, it's easier to see all matches that will be replaced. Right now, I usually just scroll up and down like crazy and try to find any selected text...which isn't that efficient. :)

2. Add companions to `buffer-replace-regex`: `buffer-search-forward-regex` and `buffer-search-backward-regex` (bound to `ctrl_shift_f` and `ctrl_shift_b`, respectively). Currently, I use `buffer-replace-regex` to search for regexes and again scroll around to find highlighted text. Again not the most efficient thing around...

3. (This one's really debatable.) If any text is selected when doing `buffer-replace*`, replace occurrences found ONLY in the region of selected text. I can see this again being useful for doing things like variable name replacements.

**Comment by kirbyfan64** [comment]
FWIW, I know about `buffer-search-word-*`, but it seems weird for `buffer-replace*` to have regex support but not `buffer-search*`.

**Comment by shalabhc** [comment]
Thanks for your suggestions!

For (1), you can use the `up` and `down` keys. Binding the same behavior also to `ctrl_f` and `ctrl_b` seems reasonable.

You make good points in (2) - I've thought about something like this also, basically to get various types of searches (incremental vs. global) for all types of syntax (regex, exact, boundary, word etc.). OTOH, this will cause an explosion of search commands so I'm divided on this. Btw, an undocumented feature of `buffer-replace` is that you can press `enter` after just typing the match regex (without typing the '/') - this jumps to the highlighted match without making any replacements. This is more useful in combination with the `up` and `down` keys, of course.

(3) is already the current behavior, if I understand you correctly. Replacements are applied only in the selected area. We should probably add highlighting in the preview for this though.

**Comment by shalabhc** [comment]
After pondering this some more, I feel a specialized regex search definitely makes sense. The problem is figuring out the right behavior. Current commands are:

Command|Matching|Region|Direction|List widget
-------|--------|------|---------|----
buffer-grep | exact + boundary | entire file | no | yes
buffer-search-forward / backward | exact | visible region | yes | no
buffer-search-word-forward / backward | exact word | visible region | yes | no

Other things to note:
* Special case: `*search-word*` commands automatically copy the word at cursor.
* There is no search that uses exact matching and displays a list
* The visible region search commands do not show a summary either (e.g. number of matches)
* No regex search

A simple solution that comes to mind is that we add `buffer-search-regex` which works like `buffer-grep` - i.e. searches the whole file and displays a summary list. This would not give a backwards incremental regex search, but maybe we can live without that? A useful feature might be if the search automatically selects the first match after the cursor position -  that might make it more like an incremental search and perhaps more useful?

This would also make the searches seem more consistent in that we would have two types of search commands - the *incremental + directional + visible region only* ones - lets call these *local* searches, and the *entire file + list widget summary* ones - lets call these *global*.

Then the question is should we also have a global exact search and global word search? Maybe we can live without these because `buffer-grep` kind of subsumes these.

In summary what I'm proposing is adding a global search command - `buffer-search-regex` - as described above. @kirbyfan64 - do you think you will be happy with this?

**Comment by kirbyfan64** [comment]
@shalabhc Sounds great!

**Comment by shalabhc** [comment]
@nilnor - any thoughts on the proposed `buffer-search-regex`?

**Comment by nilnor** [comment]
@shalabhc I wonder about the name, since it would function as the `buffer-grep-*`one, why would it not be named `buffer-grep-regex`?

**Comment by shalabhc** [comment]
@nilnor - excellent point. I was thinking about naming too. We would end up with two groups of commands - `buffer-search-*` and `buffer-grep-*` - which would makes perfect sense based on their behavior. I'll call it `buffer-grep-regex` for now. However I feel 'grep' is from unix command line only lexicon and for that reason might warrant renaming.

**Comment by shalabhc** [comment]
Check out the new branch new-buffer-greps for an implementation. Needs some specs.

**Comment by shalabhc** [comment]
Fixed with #127
